### PR TITLE
[ML] Refactor spec generation for tests

### PR DESCRIPTION
**Pull request by valeriy42**
This PR introduces a new Factory class for creating analysis specifications for test purposes to prevent code duplication.

### CODE REVIEW

**Comment by valeriy42** [comment]
retest

**Comment by valeriy42** [comment]
retest

File: include/test/CDataFrameAnalysisSpecificationFactory.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace ml {
+namespace test {
+//! \brief Collection of helping methods to create data frame analysis specifications for tests.
+class CDataFrameAnalysisSpecificationFactory {
```
Comment (tveasey) [inline include/test/CDataFrameAnalysisSpecificationFactory.h]:
```suggestion
class TEST_EXPORT CDataFrameAnalysisSpecificationFactory {
```

File: lib/test/CDataFrameAnalysisSpecificationFactory.cc [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    double alpha,
+    double lambda,
+    double gamma,
+    double softTreeDepthLimit,
+    double softTreeDepthTolerance,
+    double eta,
+    std::size_t maximumNumberTrees,
+    double featureBagFraction,
+    ml::test::CDataFrameAnalysisSpecificationFactory::TPersisterSupplier* persisterSupplier,
+    ml::test::CDataFrameAnalysisSpecificationFactory::TRestoreSearcherSupplier* restoreSearcherSupplier) {
```
Comment (tveasey) [inline lib/test/CDataFrameAnalysisSpecificationFactory.cc]:
```suggestion
    TRestoreSearcherSupplier* restoreSearcherSupplier) {
```

File: lib/test/CDataFrameAnalysisSpecificationFactory.cc [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    }
+    if (bayesianOptimisationRestarts > 0) {
+        parameters += ",\n\"bayesian_optimisation_restarts\": " +
+                      core::CStringUtils::typeToString(bayesianOptimisationRestarts);
+    }
+    parameters += "\n}";
+
+    std::string spec{api::CDataFrameAnalysisSpecificationJsonWriter::jsonString(
+        "testJob", rows, cols, memoryLimit, 1, categoricalFieldNames, true,
+        test::CTestTmpDir::tmpDir(), "ml", analysis, parameters)};
```
Comment (tveasey) [inline lib/test/CDataFrameAnalysisSpecificationFactory.cc]:
```suggestion
        CTestTmpDir::tmpDir(), "ml", analysis, parameters)};
```

File: lib/test/CDataFrameAnalysisSpecificationFactory.cc [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        parameters += (hasTrailingParameter ? ",\n" : "");
+        parameters += "\"feature_influence_threshold\": 0.0";
+        hasTrailingParameter = true;
+    }
+    parameters += (hasTrailingParameter ? "\n" : "");
+    parameters += "}\n";
+
+    std::string spec{api::CDataFrameAnalysisSpecificationJsonWriter::jsonString(
+        "testJob", rows, cols, memoryLimit, 1, {}, true,
+        test::CTestTmpDir::tmpDir(), "ml", "outlier_detection", parameters)};
```
Comment (tveasey) [inline lib/test/CDataFrameAnalysisSpecificationFactory.cc]:
```suggestion
        CTestTmpDir::tmpDir(), "ml", "outlier_detection", parameters)};
```

File: include/api/CDataFrameAnalysisSpecification.h [hunk 5]
```diff
@@ -186,12 +186,12 @@ class API_EXPORT CDataFrameAnalysisSpecification {
     //! Get pointer to the analysis runner.
     const CDataFrameAnalysisRunner* runner();
 
-private:
-    void initializeRunner(const rapidjson::Value& jsonAnalysis);
-
```
Comment (tveasey) [inline include/api/CDataFrameAnalysisSpecification.h]: The noop methods don't appear to need to be public now?

File: include/test/CDataFrameAnalysisSpecificationFactory.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+public:
+    using TStrVec = std::vector<std::string>;
+    using TDataAdderUPtr = std::unique_ptr<ml::core::CDataAdder>;
```
Comment (tveasey) [inline include/test/CDataFrameAnalysisSpecificationFactory.h]:
```suggestion
    using TDataAdderUPtr = std::unique_ptr<core::CDataAdder>;
```

File: include/test/CDataFrameAnalysisSpecificationFactory.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    using TPersisterSupplier = std::function<TDataAdderUPtr()>;
+    using TDataSearcherUPtr = std::unique_ptr<ml::core::CDataSearcher>;
```
Comment (tveasey) [inline include/test/CDataFrameAnalysisSpecificationFactory.h]:
```suggestion
    using TDataSearcherUPtr = std::unique_ptr<core::CDataSearcher>;
```

**Comment by tveasey** [comment]
One last observation you've introduced a dependency between the `test` and the `api` library. You need to update the Makefile: `LIBS:=$(LIB_ML_CORE) $(LIB_ML_MATHS) $(LIB_ML_API)`

File: include/test/CDataFrameAnalysisSpecificationFactory.h (hunk 6 above)
Comment (valeriy42) [reply include/test/CDataFrameAnalysisSpecificationFactory.h]: Done

File: include/test/CDataFrameAnalysisSpecificationFactory.h (hunk 7 above)
Comment (valeriy42) [reply include/test/CDataFrameAnalysisSpecificationFactory.h]: Done

**Pull request merged**